### PR TITLE
Extension of smartbanner.js to support OS and device type related APP links

### PR DESCRIPTION
**Pull request by Ricco1**
In on of my projects I came across the situation when app developers have created separate Apps for Phone and Tablet devices, both for iOS and Android. In the current smartbanner.js revision it is only possible to add one **Global** link, which was not suitable for my situation. So I did an extension to the current smartbanner.js revision to allow also to set the separate link for the iOS device combinations. So, as a result the current App links related metas should look like this:

```html
<meta name="smartbanner:button-url-apple" content="https://itunes.apple.com/us/genre/ios/id36?mt=8">
<meta name="smartbanner:button-url-apple-iphone" content="https://itunes.apple.com/iphonelink">
<meta name="smartbanner:button-url-apple-ipad" content="https://itunes.apple.com/ipadlink">
<meta name="smartbanner:button-url-google" content="https://play.google.com/store">
<meta name="smartbanner:button-url-google-phone" content="https://play.google.com/phonelink">
<meta name="smartbanner:button-url-google-tablet" content="https://play.google.com/tabletlink">
```

So now, if you have only one link available then it is still possible to use the old approach of setting **Global** link, while if there are several links for one OS intended for different devices you can add those by omitting the Global link meta tag. If for some reason user will leave global link and add some links to devices specific meta tags, then the latter will be ignored and the **Global** will still be used by the app.

### CODE REVIEW

**Comment by ain** [comment]
Thanks for the work, sounds interesting. I'll review the PR shortly.

File: .babelrc [hunk 1]
```diff
@@ -1,3 +1,3 @@
-{
-  "presets": [ "es2015" ]
-}
+{
+  "presets": [ "es2015" ]
+}
```
Comment (ain) [inline .babelrc]: I'm generally against touching files if nothing changes to avoid the diff. Let's keep the original one.

File: .eslintrc.json [hunk 2]
```diff
@@ -1,38 +1,38 @@
...
+    "semi": [
+      "error",
+      "always"
+    ],
+    "eqeqeq": [
+      "error",
+      "always"
+    ]
+  }
+}
```
Comment (ain) [inline .eslintrc.json]: Same diff problem here, please take a look into it. Best if the diff would have a scope or if nothing changed, no diff at all.

File: .gitignore [hunk 3]
```diff
@@ -1,31 +1,33 @@
...
+
+# Compiled binary addons (http://nodejs.org/api/addons.html)
+build/Release
+
+# Dependency directory
+# https://www.npmjs.org/doc/misc/npm-faq.html#should-i-check-my-node_modules-folder-into-git
+node_modules
+bower_components
+
+/.idea/
```
Comment (ain) [inline .gitignore]: Unclear diff. Would actually be fine if it added EOL to the file.

File: .jscsrc [hunk 4]
```diff
@@ -1,64 +1,64 @@
...
+  "disallowSpacesInAnonymousFunctionExpression": {
+      "beforeOpeningRoundBrace": true
+  },
+  "disallowSpacesInsideObjectBrackets": "all",
+  "disallowSpacesInsideArrayBrackets": "all",
+  "disallowSpacesInsideParentheses": true,
+
+  "disallowMultipleLineBreaks": true,
+  "disallowNewlineBeforeBlockStatements": true
 }
```
Comment (ain) [inline .jscsrc]: Removes EOL which is generally not a good idea. Also, coding convention change should be treated in a separate PR.

File: .scss-lint.yml [hunk 5]
```diff
@@ -1,385 +1,385 @@
...
+    enabled: true
+    identifier_list: base
+    additional_identifiers: []
+    excluded_identifiers: []
+
+  ZeroUnit:
+    enabled: true
+
+  Compass::*:
+    enabled: false
```
Comment (ain) [inline .scss-lint.yml]: Coding convention change should be treated in a separate PR if at all.

File: .tern-project [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
+    "lint": {},
+    "angular": {},
+    "requirejs": {},
+    "modules": {},
+    "es_modules": {},
+    "doc_comment": {
+      "fullDocs": true
+    }
+  }
+}
```
Comment (ain) [inline .tern-project]: Unclear diff.

File: .travis.yml [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+before_install:
+  - rvm install 2.4.0
+  - gem install bundler
+install:
+  - npm run install-deps
+script:
+  - npm run eslint
+  - npm run scsslint
+  - npm test
+  - npm run build
```
Comment (ain) [inline .travis.yml]: Unclear diff.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,33 +1,33 @@
...
+npm test
+```
+
+## Linting the SCSS files
+
+To lint an SCSS file, run:
+
+```
+npm run scsslint
 ```
```
Comment (ain) [inline CONTRIBUTING.md]: Unclear diff. Shouldn't be in scope of this PR.

File: Gemfile [hunk 9]
```diff
@@ -1,3 +1,3 @@
-source 'https://rubygems.org'
-
-gem 'scss_lint', '~> 0.48.0', require: false
+source 'https://rubygems.org'
+
+gem 'scss_lint', '~> 0.48.0', require: false
```
Comment (ain) [inline Gemfile]: Unclear diff. Shouldn't be in scope of this PR.

File: Gemfile.lock [hunk 10]
```diff
@@ -1,17 +1,18 @@
...
+
+PLATFORMS
+  ruby
+  x64-mingw32
+
+DEPENDENCIES
+  scss_lint (~> 0.48.0)
+
+BUNDLED WITH
+   1.15.1
```
Comment (ain) [inline Gemfile.lock]: Unclear diff. Shouldn't be in scope of this PR.

File: LICENSE [hunk 11]
```diff
@@ -1,674 +1,674 @@
...
+any implied license or other defenses to infringement that may
+otherwise be available to you under applicable patent law.
+
+  12. No Surrender of Others' Freedom.
+
+  If conditions are imposed on you (whether by court order, agreement or
+otherwise) that contradict the conditions of this License, they do not
+excuse you from the conditions of this License.  If you cannot convey a
+covered work so as to satisfy simultaneously your obligations under this
+License and an
```
Comment (ain) [inline LICENSE]: Licence should not be touched.

File: README.md [hunk 12]
```diff
@@ -1,98 +1,119 @@
...
+
+## Sponsors
+
+Cross-browser testing across all mobile platforms is powered by
+
+<a href="https://www.browserstack.com" title="BrowserStack"><img src="https://rawgithub.com/ain/smartbanner.js/master/browserstack.svg" height="32px" alt="BrowserStack"></a>
+
+## Licence
+
+Copyright © 2016, 2017 Ain Tohvri, contributors. Licenced under [GPL-3](https://raw.githubusercontent.com/ain/smartbanner.js/master/LICENSE).
```
Comment (ain) [inline README.md]: Diff unnecessarily too big. Please make sure your code is documented in Advanced part, the diff should only be in that section.

File: banner.tmpl [hunk 13]
```diff
@@ -1,4 +1,4 @@
-/*!
- * <%= name %> v<%= version %> <<%= homepage %>>
- * Copyright © <%= year %> <%= author %>, contributors. Licensed under <%= license %>.
- */
+/*!
+ * <%= name %> v<%= version %> <<%= homepage %>>
+ * Copyright © <%= year %> <%= author %>, contributors. Licensed under <%= license %>.
+ */
```
Comment (ain) [inline banner.tmpl]: Shouldn't be updated.

File: bower.json [hunk 14]
```diff
@@ -1,26 +1,26 @@
...
+  "ignore": [
+    "**/.*",
+    "node_modules",
+    "bower_components",
+    "test",
+    "Gemfile*",
+    "banner.tmpl",
+    "*.png"
+  ]
+}
```
Comment (ain) [inline bower.json]: Shouldn't need updated.

File: dist/index.html [hunk 15]
```diff
@@ -1,30 +1,34 @@
...
+  <script src="smartbanner.js"></script>
+</head>
+<body>
+  <h1>&nbsp;</h1>
+  <script>
+    var onLoadHandlerForConflictTesting = function() {};
+    window.onload = onLoadHandlerForConflictTesting;
+  </script>
+</body>
+</html>
```
Comment (ain) [inline dist/index.html]: Diff too big.

File: dist/smartbanner.css [hunk 16]
```diff
@@ -1,7 +1,3 @@
-/*!
- * smartbanner.js v1.5.0 <https://github.com/ain/smartbanner.js>
- * Copyright © 2017 Ain Tohvri, contributors. Licensed under GPL-3.0.
- */
```
Comment (ain) [inline dist/smartbanner.css]: Please don't update distribution in PR. That will be handled directly after merge.

File: test/spec/opionparser_spec.js [hunk 17]
```diff
@@ -1,125 +1,125 @@
...
+
+      it('expected to return empty object', function() {
+        expect(options).to.be.empty;
+      });
+
+    });
+
+  });
+
+});
```
Comment (ain) [inline test/spec/opionparser_spec.js]: Unclear diff.

File: test/spec/smartbanner_spec.js [hunk 18]
```diff
@@ -1,672 +1,984 @@
...
+          top = 0;
+        }
+        expect(top).to.eql(smartbanner.originalTop);
+        done();
+      });
+    });
+
+  });
+
+  descri
```
Comment (ain) [inline test/spec/smartbanner_spec.js]: Unclear diff.

File: src/polyfills/array/from.js [hunk 19]
```diff
@@ -1,78 +1,78 @@
...
+        }
+        k += 1;
+      }
+      // 18. Let putStatus be Put(A, "length", len, true).
+      A.length = len;
+      // 20. Return A.
+      return A;
+    };
+  }());
+}
```
Comment (ain) [inline src/polyfills/array/from.js]: Should not be changed.

File: src/detector.js [hunk 20]
```diff
@@ -1,10 +1,14 @@
 export default class Detector {
 
   static platform() {
-    if (/iPhone|iPad|iPod/i.test(window.navigator.userAgent)) {
```
Comment (ain) [inline src/detector.js]: Let's not chain this here in `platform`. I'd call for another static method, e.g. `device` which is only called if tablet or phone specific meta tag is used.

File: src/polyfills/array/includes.js [hunk 21]
```diff
@@ -1,32 +1,32 @@
...
+      currentElement = O[k];
+      if (searchElement === currentElement ||
+          (searchElement !== searchElement && currentElement !== currentElement)) { // NaN !== NaN
+        return true;
+      }
+      k++;
+    }
+    return false;
+  };
+}
```
Comment (ain) [inline src/polyfills/array/includes.js]: Should not be changed.

File: src/smartbanner.js [hunk 22]
```diff
@@ -74,6 +74,7 @@ export default class SmartBanner {
     let parser = new OptionParser();
     this.options = parser.parse();
     this.platform = Detector.platform();
+    this.platformOS = this.platform && this.platform.replace(' tablet','').replace(' phone','');
```
Comment (ain) [inline src/smartbanner.js]: As by what was discussed, this could be best named differently, e.g. `device` that'd then receive fully-parsed value from detector.

File: src/smartbanner.scss [hunk 23]
```diff
@@ -1,117 +1,117 @@
...
+    color: $blue;
+    font-size: 18px;
+    text-align: center;
+    text-decoration: none;
+
+    &__label {
+      text-align: center;
+    }
+  }
+}
```
Comment (ain) [inline src/smartbanner.scss]: Unclear diff.

File: test/spec/bakery_spec.js [hunk 24]
```diff
@@ -1,65 +1,65 @@
...
+      Bakery.bake();
+    });
+
+    it('expected to remove cookie', function() {
+      Bakery.unbake();
+      expect(Bakery.baked).to.be.false;
+    });
+
+  });
+});
```
Comment (ain) [inline test/spec/bakery_spec.js]: Should not be changed.

File: .babelrc (hunk 1 above)
Comment (Ricco1) [inline .babelrc]: Hi,
I suspect that these files that you see as touched might do something with the line breaks or some issue with my GIT, as physically none of them, except directly related to the work done, was touched.

File: .eslintrc.json (hunk 2 above)
Comment (Ricco1) [inline .eslintrc.json]: The same goes here - none of the contents should have been changed here. If you open the file you will see that for some reason GIT sees that I removed everything and pasted it back, although I've never done it. So again I assume that it is due to the global line change in the project, which I did to convert everything to LF (my default dev environment sets it as CRLF).

File: .gitignore (hunk 3 above)
Comment (Ricco1) [inline .gitignore]: Just added the .idea folder which by default is created by my dev environment, so that it doesn't get pushed to repo.

File: .jscsrc (hunk 4 above)
Comment (Ricco1) [inline .jscsrc]: Global line break change, as mentioned previously.

File: .scss-lint.yml (hunk 5 above)
Comment (Ricco1) [inline .scss-lint.yml]: Global line break change - contents are the same as previously.

File: .tern-project (hunk 6 above)
Comment (Ricco1) [inline .tern-project]: Global line break change - contents are the same as previously.

File: .travis.yml (hunk 7 above)
Comment (Ricco1) [inline .travis.yml]: Global line break change - contents are the same as previously.

File: CONTRIBUTING.md (hunk 8 above)
Comment (Ricco1) [inline CONTRIBUTING.md]: Global line break change - contents are the same as previously.

File: Gemfile (hunk 9 above)
Comment (Ricco1) [inline Gemfile]: Global line break change - contents are the same as previously.

File: Gemfile.lock (hunk 10 above)
Comment (Ricco1) [inline Gemfile.lock]: Global line break change - contents are the same as previously. + Seems that after installing new Bundle (I didn't have Ruby before), it also updated the file automatically, after running the bundle.

File: LICENSE (hunk 11 above)
Comment (Ricco1) [inline LICENSE]: Global line break change - contents are the same as previously.

File: README.md (hunk 12 above)
Comment (Ricco1) [inline README.md]: Global line break change - added only the related part to my changes.

File: banner.tmpl (hunk 13 above)
Comment (Ricco1) [inline banner.tmpl]: Global line break change - contents are the same as previously.

File: bower.json (hunk 14 above)
Comment (Ricco1) [inline bower.json]: Global line break change - contents are the same as previously.

File: dist/index.html (hunk 15 above)
Comment (Ricco1) [inline dist/index.html]: Global line break change + added the part that I implemented.

Comment (Ricco1) [inline dist/index.html]: Global line break change + my canges

File: dist/smartbanner.css (hunk 16 above)
Comment (Ricco1) [inline dist/smartbanner.css]: Had to remove '&& npm run bannerise' during the development, because I am working on windows and was getting the error during npm build during execution of this script, which is why this was automatically created without the banner during build process.

File: test/spec/opionparser_spec.js (hunk 17 above)
Comment (Ricco1) [inline test/spec/opionparser_spec.js]: Open the file to see that this involves only line change CRLF to LF

File: test/spec/smartbanner_spec.js (hunk 18 above)
Comment (Ricco1) [inline test/spec/smartbanner_spec.js]: Open the file to see the full amount of changes, which also includes CRLF to LF change (which is why you might see that all is changed)

File: src/polyfills/array/includes.js (hunk 21 above)
Comment (Ricco1) [inline src/polyfills/array/includes.js]: CRLF to LF script

File: src/detector.js (hunk 20 above)
Comment (Ricco1) [inline src/detector.js]: I'll take a look

File: src/smartbanner.js (hunk 22 above)
Comment (Ricco1) [inline src/smartbanner.js]: I'll take a look

File: src/smartbanner.scss (hunk 23 above)
Comment (Ricco1) [inline src/smartbanner.scss]: CRLF to LF

File: test/spec/bakery_spec.js (hunk 24 above)
Comment (Ricco1) [inline test/spec/bakery_spec.js]: CRLF to LF

**Comment by Ricco1** [comment]
I'll make a new pull request, so will close this one, so it does not mess up the things.
